### PR TITLE
CNS-694: validators rewards pool implementation

### PR DESCRIPTION
**Pull request by oren-lava**

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -230,16 +230,18 @@ var (
...
-		dualstakingmoduletypes.NotBondedPoolName: {authtypes.Burner, authtypes.Staking},
-		pairingmoduletypes.ModuleName:            {authtypes.Burner, authtypes.Staking},
+		authtypes.FeeCollectorName:                                       nil,
+		distrtypes.ModuleName:                                            nil,
+		stakingtypes.BondedPoolName:                                      {authtypes.Burner, authtypes.Staking},
+		stakingtypes.NotBondedPoolName:                                   {authtypes.Burner, authtypes.Staking},
+		govtypes.ModuleName:                                              {authtypes.Burner},
+		ibctransfertypes.ModuleName:                                      {authtypes.Burner},
+		subscriptionmoduletypes.ModuleName:                               {authtypes.Burner, authtypes.Staking},
+		dualstakingmoduletypes.BondedPoolName:                            {authtypes.Burner, authtypes.Staking},
```
Comment (Yaroms) [inline app/app.go:240]: remove dualstaking and pairing pools

File: config/health_examples/health_template_gen.yml
Comment (Yaroms) [inline config/health_examples/health_template_gen.yml:1]: recover

File: x/rewards/abci.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package rewards
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/x/rewards/keeper"
+)
+
+// BeginBlocker calculates the validators block rewards and transfers them to the fee collector
+func BeginBlocker(ctx sdk.Context, k keeper.Keeper) {
+	err := k.DistributeBlockReward(ctx)
```
Comment (Yaroms) [inline x/rewards/abci.go:10]: why does this return an error?

Comment (oren-lava) [reply x/rewards/abci.go]: done. see https://github.com/lavanet/lava/commit/3c190e65adf95d112a07287483dce09123539c1a

File: app/app.go (hunk 1 above)
Comment (oren-lava) [reply app/app.go]: done. see https://github.com/lavanet/lava/commit/3c190e65adf95d112a07287483dce09123539c1a

**Review by Yaroms — CHANGES_REQUESTED** [review]

File: testutil/keeper/mock_keepers.go [hunk 3]
```diff
@@ -161,11 +168,13 @@ func (k mockBankKeeper) SetBalance(ctx sdk.Context, addr sdk.AccAddress, amounts
 func (k mockBankKeeper) AddToBalance(addr sdk.AccAddress, amounts sdk.Coins) error {
 	if _, ok := balance[addr.String()]; ok {
 		for _, coin := range amounts {
+			fmt.Printf("before balance[addr.String()].String(): %v\n", balance[addr.String()].String())
 			balance[addr.String()] = balance[addr.String()].Add(coin)
 		}
 	} else {
 		balance[addr.String()] = amounts
 	}
+	fmt.Printf("after balance[addr.String()].String(): %v\n", balance[addr.String()].String())
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply testutil/keeper/mock_keepers.go]: done. see https://github.com/lavanet/lava/pull/1040/commits/f0e86a906efd6cfc82a0d78236f5d50825fc9b84

File: testutil/keeper/mock_keepers.go [hunk 4]
```diff
@@ -264,3 +273,10 @@ func (b *MockBlockStore) LoadBlockMetaByHash(hash []byte) *tenderminttypes.Block
 func (b *MockBlockStore) DeleteLatestBlock() error {
 	return nil
 }
+
+// rewards pool mock
+type mockRewardsPool struct{}
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply testutil/keeper/mock_keepers.go]: done. see https://github.com/lavanet/lava/commit/be375085865201c4b5b5ace62a2e7c9ace9106b1

File: x/rewards/keeper/rewards.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+//  1. burns the current token in the validators distribution pool by the burn rate
+//  2. transfers the monthly tokens quota from the validators allocation pool to the validators distribution pool
+//  3. opens a new timer for the next month (and encodes the expiry block and months left to allocation pool in it)
+//
+// for providers:
+// TBD
+func (k Keeper) RefillRewardsPools(ctx sdk.Context, _ []byte, data []byte) {
+	// get the months left for the allocation pools
+	var monthsLeft uint64
+	if len(data) == 0 {
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/rewards.go:62]: it's only 0 on genesis (see `x/rewards/genesis.go` line 16)

File: x/rewards/keeper/rewards.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	blocksToNextTimerExpiry := ((durationUntilNextMonth / int64(blockCreationTime)) * 105 / 100) + ctx.BlockHeight()
+
+	// update the months left of the allocation pool and encode it
+	monthsLeftBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(monthsLeftBytes, monthsLeft)
+
+	// open a new timer for next month
+	blocksToNextTimerExpirybytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(blocksToNextTimerExpirybytes, uint64(blocksToNextTimerExpiry))
+	k.refillRewardsPoolTS.AddTimerByBlockTime(ctx, uint64(nextMonth), blocksToNextTimerExpirybytes, monthsLeftBytes)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/rewards.go:87]: talked about it and I won't change anything.
Basically, @Yaroms suggested to lose the timer all together and keep in a KV store the month's next timestamp, blocks to next timer expiry and month left. we decided it's a bit overkill just to make the code "less weird". @omerlavanet what do you think?

File: x/rewards/keeper/rewards.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+func (k Keeper) DistributeBlockReward(ctx sdk.Context) error {
+	// get params for validator rewards calculation
+	bondedTargetFactor := k.bondedTargetFactor(ctx)
+	blocksToNextTimerExpiry := k.BlocksToNextTimerExpiry(ctx)
+
+	// get validator distribution pool balance
+	distributionPoolBalance := k.TotalPoolTokens(ctx, types.ValidatorsRewardsDistributionPoolName)
+
+	// validators bonus rewards = (distributionPoolBalance * bondedTargetFactor) / blocksToNextTimerExpiry
+	validatorsRewards := bondedTargetFactor.MulInt(distributionPoolBalance).TruncateInt().QuoRaw(blocksToNextTimerExpiry)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/rewards.go]: done. see https://github.com/lavanet/lava/commit/9976928e28974480145d8ed17f7507eb07fda7a2

File: x/rewards/keeper/rewards.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+	"math"
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/utils"
+	"github.com/lavanet/lava/x/rewards/types"
+	timerstoretypes "github.com/lavanet/lava/x/timerstore/types"
+)
+
+func (k Keeper) DistributeBlockReward(ctx sdk.Context) error {
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/rewards.go]: done. see https://github.com/lavanet/lava/commit/9976928e28974480145d8ed17f7507eb07fda7a2

File: x/rewards/keeper/rewards.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+	err := k.BurnPoolTokens(ctx, distributionPool, tokensToBurn)
+	if err != nil {
+		utils.LavaFormatError("critical - could not burn distribution pool tokens", err,
+			utils.Attribute{Key: "distribution_pool", Value: string(distributionPool)},
+			utils.Attribute{Key: "tokens_to_burn", Value: tokensToBurn.String()},
+		)
+	}
+
+	// transfer the new monthly quota (if allocation pool is expired, rewards=0)
+	monthlyQuota := sdk.Coins{sdk.Coin{Denom: k.stakingKeeper.BondDenom(ctx), Amount: sdk.ZeroInt()}}
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/rewards.go]: done. see https://github.com/lavanet/lava/commit/9976928e28974480145d8ed17f7507eb07fda7a2

File: x/rewards/keeper/rewards.go [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+	allocPoolBalance := k.TotalPoolTokens(ctx, allocationPool)
+	if monthsLeft != 0 && !allocPoolBalance.IsZero() {
+		monthlyQuota[0] = monthlyQuota[0].AddAmount(allocPoolBalance.QuoRaw(int64(monthsLeft)))
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/rewards.go]: done. see https://github.com/lavanet/lava/commit/9976928e28974480145d8ed17f7507eb07fda7a2

File: x/rewards/keeper/rewards.go [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+			utils.Attribute{Key: "distribution_pool_balance", Value: distributionPoolBalance.String()},
+			utils.Attribute{Key: "blocks_to_next_timer_expiry", Value: strconv.FormatInt(blocksToNextTimerExpiry, 10)},
+		)
+		return nil
+	}
+
+	coins := sdk.NewCoins(sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), validatorsRewards))
+
+	// distribute rewards to validators (same as Cosmos mint module)
+	err := k.addCollectedFees(ctx, coins)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/rewards.go]: done. see https://github.com/lavanet/lava/commit/9976928e28974480145d8ed17f7507eb07fda7a2

File: x/rewards/abci.go [hunk 12]
```diff
@@ -7,8 +7,5 @@ import (
 
 // BeginBlocker calculates the validators block rewards and transfers them to the fee collector
 func BeginBlocker(ctx sdk.Context, k keeper.Keeper) {
-	err := k.DistributeBlockReward(ctx)
-	if err != nil {
-		panic(err)
-	}
+	k.DistributeBlockReward(ctx)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/abci.go]: done. see https://github.com/lavanet/lava/commit/9976928e28974480145d8ed17f7507eb07fda7a2

**Review by omerlavanet — CHANGES_REQUESTED** [review]

File: x/rewards/keeper/keeper.go [hunk 13]
```diff
@@ -58,13 +66,30 @@ func NewKeeper(
...
+
+	return &keeper
 }
 
 func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
 }
 
-func (k Keeper) BondedTargetFactor(ctx sdk.Context) cosmosMath.LegacyDec {
+// redeclaring BeginBlock for testing (this is not called outside of unit tests)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/keeper.go:85]: outdated. Not there anymore

File: x/rewards/keeper/rewards.go [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+	if monthsLeft > 0 {
+		monthsLeft -= 1
+	}
+
+	// calculate the block in which the timer will expire (+5% for errors)
+	nextMonth := utils.NextMonth(ctx.BlockTime()).UTC().Unix()
+	durationUntilNextMonth := nextMonth - ctx.BlockTime().UTC().Unix()
+	blockCreationTime := k.downtimeKeeper.GetParams(ctx).DowntimeDuration.Seconds()
+	blocksToNextTimerExpiry := ((durationUntilNextMonth / int64(blockCreationTime)) * 105 / 100) + ctx.BlockHeight()
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/rewards.go]: done. see https://github.com/lavanet/lava/commit/8c0f9f80f1660953e8ef0e877143479d79c7b738

File: x/rewards/genesis.go [hunk 15]
```diff
@@ -4,20 +4,23 @@ import (
...
+	timerstoretypes "github.com/lavanet/lava/x/timerstore/types"
 )
 
 // InitGenesis initializes the capability module's state from a provided genesis
 // state.
 func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState) {
 	// this line is used by starport scaffolding # genesis/module/init
 	k.SetParams(ctx, genState.Params)
+	k.InitRewardsRefillTS(ctx, *timerstoretypes.DefaultGenesis())
+	k.RefillRewardsPools(ctx, nil, nil)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/genesis.go]: done. see https://github.com/lavanet/lava/commit/c5b5133bd4a64dbb9f8939cf60ccb63fac855510

File: x/rewards/keeper/rewards.go [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+	// update the months left of the allocation pool and encode it
+	monthsLeftBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(monthsLeftBytes, monthsLeft)
+
+	// open a new timer for next month
+	blocksToNextTimerExpirybytes := make([]byte, 8)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/rewards.go]: done. see https://github.com/lavanet/lava/commit/3bc57748a21eeeb11cbd9d7b7058e1e9ebb539e3

File: x/rewards/keeper/rewards.go [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+// BlocksToNextTimerExpiry extracts the timer's expiry block from the timer's subkey
+// and returns the amount of blocks remaining (according to the current block height)
+func (k Keeper) BlocksToNextTimerExpiry(ctx sdk.Context) int64 {
+	keys, _, _ := k.refillRewardsPoolTS.GetFrontTimers(ctx, timerstoretypes.BlockTime)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/rewards.go]: done. see https://github.com/lavanet/lava/commit/3bc57748a21eeeb11cbd9d7b7058e1e9ebb539e3

File: x/rewards/keeper/rewards.go [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+	if len(keys) == 0 {
+		// something is wrong, don't panic but make validators rewards 0
+		utils.LavaFormatError("could not get blocks to next timer expiry", fmt.Errorf("no timers found"))
+		return math.MaxInt64
+	}
+	return int64(binary.BigEndian.Uint64(keys[0])) - ctx.BlockHeight()
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/rewards.go]: done. see https://github.com/lavanet/lava/commit/3bc57748a21eeeb11cbd9d7b7058e1e9ebb539e3

File: x/rewards/keeper/pool_test.go [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+	err = ts.TxProposalChangeParam(types.ModuleName, paramKey, paramVal)
+	require.Nil(t, err)
+
+	// compare the new block reward to refBlockReward
+	params := ts.Keepers.Rewards.GetParams(ts.Ctx)
+	maxBonded := params.MaxBondedTarget
+	bonded := ts.Keepers.StakingKeeper.BondedRatio(ts.Ctx)
+	lowFactor := params.LowFactor
+	e1 := maxBonded.Sub(bonded).Quo(maxBonded)
+	e2 := bonded.Quo(maxBonded)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/pool_test.go]: changed the whole test to be using hard-coded values. see https://github.com/lavanet/lava/commit/a1586fded7e08e27105b7dbe491716645583d344

File: x/rewards/keeper/pool_test.go [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+	// Advance 3 blocks and check again
+	ts.AdvanceBlocks(3)
+	expectedTimeToExpiry := secondsInAMonth - 3*blockCreationTime
+	timeToExpiry = ts.Keepers.Rewards.TimeToNextTimerExpiry(ts.Ctx)
+	require.Equal(t, expectedTimeToExpiry, timeToExpiry)
+
+	expectedBlocksToExpiry := blocksInAMonth - 3
+	blocksToExpiry = ts.Keepers.Rewards.BlocksToNextTimerExpiry(ts.Ctx)
+	require.Equal(t, expectedBlocksToExpiry, blocksToExpiry)
+}
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/pool_test.go:275]: we test the bondedTargetFactor is calculated as expected (and changes over blocks advancement) in the test "TestValidatorBlockRewards". see https://github.com/lavanet/lava/commit/a1586fded7e08e27105b7dbe491716645583d344

for bondedTargetFactor edge cases test, see https://github.com/lavanet/lava/commit/69c0294de4f4958b90b23b7aca84bc0f66bcc6a2

File: x/rewards/keeper/pool_test.go [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+	// the last three iterations will be after the allocation pool's funds are depleted
+	var feeCollectorFinalBalance int64
+	for i := 0; i < int(lifetime+2); i++ {
+		// to see why these 3 are called, see general note 2
+		ts.AdvanceMonths(1)
+		ts.AdvanceBlock()
+		testkeeper.EndBlock(ts.Ctx, ts.Keepers)
+
+		// check the allocation pool transfers the expected monthly quota each month
+		if i >= 47 {
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/pool_test.go]: I tried using a playbook but it made the code messy. I rewrote the test to separate the different periods of the allocation pool lifetime. Hope it's better now. see https://github.com/lavanet/lava/commit/8cad9b41a0ea1e7fa1842f9bb980f38c00e835ed

File: x/rewards/keeper/pool_test.go [hunk 22]
```diff
@@ -0,0 +1,259 @@
...
+			if feeCollectorFinalBalance == 0 {
+				feeCollectorFinalBalance = ts.GetBalance(ts.feeCollector())
+			} else {
+				// fee collector balance should not increase (rewards = 0)
+				balance := ts.GetBalance(ts.feeCollector())
+				require.Equal(t, feeCollectorFinalBalance, balance)
+			}
+		} else {
+			// adding 1 because setup did the first month
+			expectedMonthlyQuota = currentAllocPoolBalance.Int64() / (lifetime - (int64(i) + 1))
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/pool_test.go]: done. see https://github.com/lavanet/lava/commit/8cad9b41a0ea1e7fa1842f9bb980f38c00e835ed

File: x/rewards/keeper/pool_test.go [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+		}
+		prevAllocPoolBalance := currentAllocPoolBalance
+		currentAllocPoolBalance = ts.Keepers.Rewards.TotalPoolTokens(ts.Ctx, types.ValidatorsRewardsAllocationPoolName)
+		require.Equal(t, expectedMonthlyQuota, prevAllocPoolBalance.Sub(currentAllocPoolBalance).Int64())
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/pool_test.go]: done. see https://github.com/lavanet/lava/commit/8cad9b41a0ea1e7fa1842f9bb980f38c00e835ed

File: x/rewards/keeper/pool_test.go [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+
+	// Advance 3 blocks and check again
+	ts.AdvanceBlocks(3)
+	expectedTimeToExpiry := secondsInAMonth - 3*blockCreationTime
+	timeToExpiry = ts.Keepers.Rewards.TimeToNextTimerExpiry(ts.Ctx)
+	require.Equal(t, expectedTimeToExpiry, timeToExpiry)
+
+	expectedBlocksToExpiry := blocksInAMonth - 3
+	blocksToExpiry = ts.Keepers.Rewards.BlocksToNextTimerExpiry(ts.Ctx)
+	require.Equal(t, expectedBlocksToExpiry, blocksToExpiry)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/rewards/keeper/pool_test.go:295]: done. see https://github.com/lavanet/lava/commit/7f12c19eb774b854431dff0973bad26cc7c09ba0

**Review by Yaroms — APPROVED** [review]
